### ISSUE TITLE
Now sgp4 function is taking a lot more time than 1.3.0 version

### ISSUE THREAD

**Issue by adrianhescalante** [shashwatak/satellite-js#45]
After updating from version 1.3.0 to 2.0.2, now the sgp4 function is taking like 100 times the previous time to calculate the propagation. After profiling my code all points out that sgp4 is the problematic function. For now I will stay in 1.3.0, but I just want to point it out.

Is it normal or maybe just a bug?

**Comment by terekcampbell** [comment]
+1
I just upgraded from 1.3.0 to 2.0.2 and noticed a similar significant slowdown. I didn't run a profiler to verify it's coming from that function, but a spot check implies it is. Until it's resolved I've got to stick with 1.3.0

**Comment by thkruz** [comment]
Few observations:

- sgp4 function declares rec and then uses _extends to do something unlike 1.3.
- - I replaced that line with var rec = satrec.
- sgp4 function returns satrec in 2.0.2
- - I set it to only return postion and velocity.

In my implementation:
2.0.2 takes 757ms to propagate the whole catalogue
2.0.2 without the above code takes 175ms (no other changes)
1.3 takes 58ms.

Does anyone know the purpose of _extends?

**Comment by ezze** [comment]
Hi, guys! Thanks for reporting this.

@thkruz, the purpose of `extends` is to clone the original `satrec` argument:

https://github.com/shashwatak/satellite-js/blob/develop/src/propagation/sgp4.js#L101

Otherwise some fields of `satrec` input object will be modified. I'm not sure whether this is an allowed behavior and will not break anything else. Before reverting it back I would try to replace spread operator by `Object.assign` or some simple function for cloning the `satrec`. It's hard to believe that cloning the object takes ~ half of a second.

**Comment by ezze** [comment]
I've checked that in 1.3.0 `satrec` is really not returned as the result of `sgp4` function and, probably, that all modified properties of this object is for internal usage only (excepting `satrec.error`).

**Comment by ezze** [comment]
@adrianhescalante, @terekcampbell, @thkruz, guys, please check out version 2.0.3. Hopefully, it's fixed now.

`satrec` object is not cloned in `sgp4` and `sgp4init` functions as it was in 1.3.0. I also had to add `eslint-disable no-param-reassign` instructions to satisfy ESLint.

**Comment by thkruz** [comment]
Fairly similar results between 2.0.3 and 1.3. Well done.

**Comment by thkruz** [comment]
So I spent the day doing a lot more benchmarking and stand corrected. I did a loop of 20x17000 sgp4 calculations on two random satellites (one near earth and one deep space). Results:

- 1.3: 32.35ms
- 2.0.3: 30.75ms
- 2.0.3 Repackaged Not ES6 for WebWorker: 29.75ms

Now when I take that repackaged version of 2.0.3 and put it into my actual webworker it takes 55ms to iterate the catalogue and 32ms with 1.3 (all else being equal). The only function called with any noticeable change in performance is sgp4().

I downloaded various versions of satellite.js and substituted them in to see when the change happened, and it appears to start with 1.4 during the migration to ES6. What stands out as being different between 1.2 and 1.4 is the almond library; it seems to have really given this library a performance boost when used in a webworker.

Is it possible that block scoped variables are creating more garbage and the garbage collector is what is making the difference?

For most applications, I think 2.0.3 is negligibly slower (or even possibly faster), but in my case of wanting the whole catalogue propagated near non-stop in a webworker, 25 ms is extremely noticeable.
